### PR TITLE
Refactoring of internal/value package implementation with zero allocations

### PR DESCRIPTION
**Pull request by asmyasnikov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

### CODE REVIEW

File: internal/value/exp/allocator/pool_go1.18.go [hunk 1]
```diff
@@ -10,37 +10,41 @@ import (
...
-	uint32Pool        pool[Ydb.Value_Uint32Value]
-	low128Pool        pool[Ydb.Value_Low_128]
-	int64Pool         pool[Ydb.Value_Int64Value]
-	uint64Pool        pool[Ydb.Value_Uint64Value]
-	floatPool         pool[Ydb.Value_FloatValue]
-	doublePool        pool[Ydb.Value_DoubleValue]
-	nestedPool        pool[Ydb.Value_NestedValue]
-	pairPool          pool[Ydb.ValuePair]
+	allocatorsPool         pool[Allocator]
+	valuePool              pool[Ydb.Value]
```
[the remark this answers is not in the document]
Comment (positiveviking) [reply internal/value/exp/allocator/pool_go1.18.go:14]: this solution requires mutes - this is weak point in high loaded apps
generate wrappers over sync.Pools is more reasonable idea

File: internal/value/exp/type_empty_list.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+type emptyListType struct{}
+
+func (v emptyListType) toString(buffer *bytes.Buffer) {
+	buffer.WriteString("List<>")
+}
+
+func (v emptyListType) String() string {
+	var buf bytes.Buffer
```
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/type_empty_list.go]: ack

File: internal/value/exp/type_decimal.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (v *DecimalType) String() string {
+	var buf bytes.Buffer
+	v.toString(&buf)
+	return buf.String()
+}
+
+func (v *DecimalType) equalsTo(rhs T) bool {
+	vv, ok := rhs.(*DecimalType)
```
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/type_decimal.go]: ack

File: internal/value/exp/type_decimal.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/value/exp/allocator"
+)
+
+type DecimalType struct {
+	Precision uint32
+	Scale     uint32
+}
+
+func (v *DecimalType) toString(buffer *bytes.Buffer) {
+	buffer.WriteString("Decimal(")
```
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/type_decimal.go]: ack

File: internal/value/exp/allocator/pool_go1.18.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/allocator/pool_go1.18.go]: decline

**Comment by asmyasnikov** [comment]
Benchmarking allocations:
| `internal/value` implementation | `ns/op` | `B/op` | `allocs/op` |
|---------------------------------|--------|--------|------------|
| Current | 14122 | 11304 | 195 |
| Proposal | 8249 | 0 | 0 |

File: internal/value/exp/type_decimal.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"bytes"
+	"strconv"
+
+	"github.com/ydb-platform/ydb-go-genproto/protos/Ydb"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/value/exp/allocator"
+)
+
+type DecimalType struct {
```
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/type_decimal.go]: ack

File: internal/value/exp/type.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ydb-platform/ydb-go-genproto/protos/Ydb"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/value/exp/allocator"
+)
+
+type T interface {
```
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/type.go]: ack

File: internal/value/exp/value.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	"encoding/binary"
+	"fmt"
+	"strconv"
+
+	"github.com/ydb-platform/ydb-go-genproto/protos/Ydb"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/value/exp/allocator"
+)
+
+type V interface {
```
[the remark this answers is not in the document]
Comment (asmyasnikov) [reply internal/value/exp/value.go]: ack

**Pull request merged**